creusot/src/backend/program.rs

```diff
@@ -1081,23 +1087,29 @@ impl<'tcx> Statement<'tcx> {
                 istmts.extend(assign);
                 istmts
             }
-            Statement::Resolve { did, subst, pl, pat } => {
+            Statement::Resolve { did, subst, pl } => {
                 lower.ctx.translate(did);
                 let mut istmts = Vec::new();
 
                 let rp = Exp::qvar(lower.names.value(did, subst));
                 let loc = pl.local;
 
-                let mut exp = rp.app_to(pl.as_rplace(lower, &mut istmts));
-                if let Some(pat) = pat {
-                    exp = Exp::Match(
+                let bound = lower.fresh_sym_from("inv_asrt");
+
+                let pat = pattern_of_place(lower.ctx.tcx, lower.locals, pl, bound);
+
+                let pat = lower_pat(lower.ctx, lower.names, &pat);
+                let exp = if let Pattern::VarP(_) = pat {
+                    rp.app_to(Exp::var(util::ident_of(loc)))
+                } else {
+                    Exp::Match(
```

Comment:
I kept the optimization present before which avoided introducing a match on a wildcard (seems reasonable).